src/scheduler/gc_work.rs

```diff
@@ -245,24 +247,148 @@ impl<VM: VMBinding> GCWork<VM> for EndOfGC {
+        let mut process_edges_work = E::new(vec![], false, mmtk);
+        // FIXME: This line allows us to omit the borrowing lifetime of worker.
+        // We should refactor ProcessEdgesWork so that it uses `worker` locally, not as a member.
+        process_edges_work.set_worker(worker);
+
+        // Cretae the tracer.
+        let mut tracer = ProcessEdgesWorkTracer { process_edges_work };
+
+        // The caller can use the tracer here.
+        let result = func(&mut tracer);
+
+        // Flush the queued nodes.
+        let ProcessEdgesWorkTracer {
+            mut process_edges_work,
+        } = tracer;
+        let next_nodes = process_edges_work.pop_nodes();
+        if !next_nodes.is_empty() {
+            // Divide the resulting nodes into appropriately sized packets.
+            let work_packets = next_nodes
+                .chunks(E::CAPACITY)
+                .map(|chunk| {
+                    Box::new(process_edges_work.create_scan_work(chunk.into(), false)) as _
+                })
+                .collect::<Vec<_>>();
+            worker.scheduler().work_buckets[self.stage].bulk_add(work_packets);
+        }
+
+        result
+    }
+}
+
+/// Delegate to the VM binding for weak reference processing.
 ///
 /// Some VMs (e.g. v8) do not have a Java-like global weak reference storage, and the
 /// processing of those weakrefs may be more complex. For such case, we delegate to the
 /// VM binding to process weak references.
-#[derive(Default)]
-pub struct VMProcessWeakRefs<E: ProcessEdgesWork>(PhantomData<E>);
+///
+/// NOTE: This will replace `{Soft,Weak,Phantom}RefProcessing` and `Finalization` in the future.
+pub struct VMProcessWeakRefs<E: ProcessEdgesWork> {
+    phantom_data: PhantomData<E>,
+}
 
 impl<E: ProcessEdgesWork> VMProcessWeakRefs<E> {
     pub fn new() -> Self {
-        Self(PhantomData)
+        Self {
+            phantom_data: PhantomData,
+        }
     }
 }
 
 impl<E: ProcessEdgesWork> GCWork<E::VM> for VMProcessWeakRefs<E> {
     fn do_work(&mut self, worker: &mut GCWorker<E::VM>, _mmtk: &'static MMTK<E::VM>) {
-        trace!("ProcessWeakRefs");
-        <E::VM as VMBinding>::VMCollection::process_weak_refs(worker); // TODO: Pass a factory/callback to decide what work packet to create.
+        trace!("VMProcessWeakRefs");
+
+        let stage = WorkBucketStage::VMRefClosure;
+
+        let need_to_repeat = {
+            let tracer_factory = ProcessEdgesWorkTracerContext::<E> {
+                stage,
+                phantom_data: PhantomData,
+            };
+            <E::VM as VMBinding>::VMScanning::process_weak_refs(worker, tracer_factory)
+        };
+
+        if need_to_repeat {
+            // Schedule Self as the new sentinel so we'll call `process_weak_refs` again after the
+            // current transitive closure.
+            let new_self = Box::new(Self::new());
+
+            worker.scheduler().work_buckets[stage].set_sentinel(new_self);
```

Comment:
If I understand correctly, this is used for performing weak-ref tracing in a loop. You can use the sentinel packet to start the transitive closure again and again until reaching a steady state.

However, I think OpenJDK may also use the sentinel packet as a "state transition" to start a new phase for a different type of weak-ref processing. For example, OpenJDK does soft-ref processing first (which will start a round of closure), and then phantom processing later (and another round of closure).  If the VM is able to set a different sentinel packet itself other than `new_self `, after soft-ref processing, it can set a phantom sentinel packet to start phantom processing. Is it doable?

Reply:
To implement state transition, we can embed a state machine in the weak reference processor, and multiplex the operations for different states through the single `Scanning::process_weak_refs` API. See:  https://github.com/wks/mmtk-openjdk/blob/gen-weakref-api/mmtk/src/weak_processor/mod.rs#L61

It is possible to add a way to specify the next work packet to be scheduled as the "sentinel" packet.  But that may be more complicated (as it has to expose details such as "sentinel") than simply returning `true` and letting the function be called again.